### PR TITLE
MINOR: Only require re-join if the metadata snapshot's topics info has changed

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @hachikuji for review.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 1]
```diff
@@ -808,6 +808,31 @@ public boolean equals(Object o) {
...
+                                && !that.partitionsPerTopic.get(topic).equals(this.partitionsPerTopic.get(topic))) {
+
+                            return true;
+                        }
+                    } else {
+                        return true;
+                    }
+                }
+
+                return false;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:833]: Hmm, are you doing anything different than `partitionsPerTopic == null : that.partitionsPerTopic == null : partitionsPerTopic.equals(that.partitionsPerTopic)`?

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:833]: Ah, you ignore the topics that exist in `that.partitionsPerTopic`, but not in `partitionsPerTopic`?

**Comment by guozhangwang** [comment]
Discussed with @hachikuji offline and we think this issue deserves a better solution, while this PR itself is not perfectly resolving it while introducing other issues. So closing it for now.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 1 above)
Comment (eribeiro) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:833]: Hey @guozhangwang, what do you think about changing the structure of this method as below. My comments are inserted below:

```
public boolean topicsChanged(MetadataSnapshot that) {

       if (partitionsPerTopic == null)
           return that.partitionsPerTopic == null;

      // if that.partitionsPerTopic == null && partitionsPerTopic != null
      // then we have NPE below, right? Is that possible?

       if (partitionsPerTopic.size() != that.partitionsPerTopic.size())
           return true;

       // check if Sets are equal, as size (number of entries) is equal (per above) then it must be the same 
       if (!that.partitionsPerTopic.keySet().containsAll(partitionsPerTopic.keySet())
           return false;

       for (Entry<String,Integer> e: partitionsPerTopic.entrySet()) {
             String topic = e.getKey();
             Integer partitions = e.getValue();
             Integer thatPartitions =  that.partitionsPerTopic.get(topic);
             if (partitions  != null &&
                 // THE CHECK BELOW IS NEEDED? partitions.equals(null) is false, right?
                 thatPartitions != null && 
                 !partitions.equals(thatPartitions)) {
                 return true;
             }
        }

        return false;
}
```

The logic doesn't change (I hope so! haha), but the changes make it little bit easier to read, imho. ;)

Comment (eribeiro) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:833]: Sorry for the dumb question, but I am curious if

```
     if (partitionsPerTopic == null)
           return that.partitionsPerTopic == null;
```

Wouldn't be

```
       if ((partitionsPerTopic == null && that.partitionsPerTopic != null) ||
           (partitionsPerTopic != null && that.partitionsPerTopic == null))
           return true;
```

**Comment by eribeiro** [comment]
LOL, sorry. 😄  Don't remember how landed on this PR, but didn't see you had closed the issue. Excuse me.

**Pull request closed without merging**